calicoctl/commands/ipam/show.go

```diff
@@ -21,83 +21,135 @@ import (
 	"os"
 	"strings"
 
+	"github.com/apex/log"
 	"github.com/olekukonko/tablewriter"
 
 	"github.com/projectcalico/calicoctl/calicoctl/commands/constants"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	"github.com/projectcalico/libcalico-go/lib/clientv3"
 	cerrors "github.com/projectcalico/libcalico-go/lib/errors"
 	"github.com/projectcalico/libcalico-go/lib/ipam"
+	"github.com/projectcalico/libcalico-go/lib/net"
+	"github.com/projectcalico/libcalico-go/lib/options"
 
 	docopt "github.com/docopt/docopt-go"
 
 	"github.com/projectcalico/calicoctl/calicoctl/commands/argutils"
 	"github.com/projectcalico/calicoctl/calicoctl/commands/clientmgr"
+	bapi "github.com/projectcalico/libcalico-go/lib/backend/api"
 )
 
-// IPAM takes keyword with an IP address then calls the subcommands.
-func Show(args []string) error {
-	doc := constants.DatastoreIntro + `Usage:
-  calicoctl ipam show [--ip=<IP> | --show-blocks] [--config=<CONFIG>]
-
-Options:
-  -h --help             Show this screen.
-     --ip=<IP>          Report whether this specific IP address is in use.
-     --show-blocks      Show detailed information for IP blocks as well as pools.
-  -c --config=<CONFIG>  Path to the file containing connection configuration in
-                        YAML or JSON format.
-                        [default: ` + constants.DefaultConfigPath + `]
+func getPoolDetails(ctx context.Context, ippoolClient clientv3.IPPoolInterface, bc bapi.Client) ([]*ipam.PoolDetails, error) {
+	var details []*ipam.PoolDetails
 
-Description:
-  The ipam show command prints information about a given IP address, or about
-  overall IP usage.
-`
-	parsedArgs, err := docopt.Parse(doc, args, true, "", false, false)
+	allPools, err := ippoolClient.List(ctx, options.ListOptions{})
 	if err != nil {
-		return fmt.Errorf("Invalid option: 'calicoctl %s'. Use flag '--help' to read about a specific subcommand.", strings.Join(args, " "))
+		return nil, err
 	}
-	if len(parsedArgs) == 0 {
-		return nil
+
+	for _, pool := range allPools.Items {
```

Comment:
Could we simplify our data structures here?

Instead of a nested structure, with pools at the top level and then another list below, what about something like this?

```go
type BorrowedIP struct {
  addr string
  borrowingNode string
  block string
  blockOwner string
  allocationType string
  allocatedTo string
  pool string
}
```

And then we just use a `[]BorrowedIP` to populate the table? This is all assuming we even need IP pool in the table output. Is there a requirement for that?

Reply:
There is no requirement for ippool so used your suggested structure removing pool.